File: shared/web-sdk/src/connectors/assistant-audio-response-connector.ts
import type { Connector, SentientSDKInternal } from "../connector-types.ts";
import { createLogger } from "../logger.ts";

const log = createLogger(["sentient", "sdk", "connectors", "assistant-audio-response"]);

/** Wire encoding of the outbound TTS byte stream, read off `turn.audio.start`. */
export interface AssistantAudioFormat {
  readonly encoding: "opus" | "pcm";
  readonly sampleRate: number;
}

export interface AssistantAudioResponseConfig {
  /** One outbound audio frame (payload bytes, header already peeled by the
   *  router). `turnId` comes from the most recent `turn.audio.start` — the
   *  binary frame itself carries no turn id. */
  onAudioFrame?: (frame: Uint8Array, turnId: string) => void;
  /** A new audio stream started. `format` is omitted if the gateway did not
   *  send both fields — the connector never invents a codec or a rate. */
  onAudioStart?: (turnId: string, format?: AssistantAudioFormat) => void;
  /** The audio stream for this turn completed normally. */
  onAudioDone?: (turnId: string) => void;
  /** Gateway signalled a mid-stream playback halt. Drop queued audio NOW.
   *  Emitted ONLY on barge-in (mic onset) or interrupt (UI Stop) — a new
   *  turnId never produces this frame (spec §4.7 / §7.2). */
  onPlaybackStop?: (reason: "barge-in" | "interrupt", turnId: string) => void;
}

// ---------------------------------------------------------------------------
// AssistantAudioResponseConnector — receives assistant TTS audio.
//
// Capability: "audio.output"
// Direction: output
//
// Receives (2.0 wire contract):
//   turn.audio.start  → open a stream, remember its turnId
//   binary frames     → payload bytes attributed to the open stream
//   turn.audio.done   → close the stream
//   playback.stop     → user-initiated halt; suppress frames until the next start
//
// FRAME ATTRIBUTION CONTRACT: outbound binary audio carries no turn id, so
// the gateway emits one turn's audio at a time, delimited by
// turn.audio.start / turn.audio.done. Overlapping streams would make frames
// unattributable — the connector WARNs loudly rather than silently
// mislabelling the tail of the previous turn. Sequential EMISSION does not
// mean sequential PLAYBACK: the next turn's frames still arrive while the
// previous turn's audio is buffered in the output pipeline, which is exactly
// what TurnAudioQueue exists to serialize (§7.2).
// ---------------------------------------------------------------------------

export class AssistantAudioResponseConnector implements Connector {
  readonly capability = "audio.output";
  readonly kind = "output" as const;

  private readonly config: AssistantAudioResponseConfig;
  private unsubs: (() => void)[] = [];
  private isReceiving = false;
  private isCancelled = false;
  /** turnId the currently arriving binary frames belong to. */
  private activeTurnId = "";

  constructor(config: AssistantAudioResponseConfig = {}) {
    this.config = config;
  }

  attach(sdk: SentientSDKInternal): void {
    this.isCancelled = false;

    this.unsubs.push(
      sdk.onMessage("turn.audio.start", (msg: unknown) => {
        const m = msg as { turnId?: string; encoding?: string; sampleRate?: number };
        if (this.isReceiving) {
          log.warn("audio-start-while-receiving", {
            reason: "overlapping turn audio streams — frames cannot be attributed",
            previousTurnId: this.activeTurnId,
            turnId: m.turnId ?? "",
          });
        }
        this.activeTurnId = m.turnId ?? "";
        this.isReceiving = true;
        this.isCancelled = false;
        const { encoding, sampleRate } = m;
        // Annotated so the narrowed literal union survives object-literal
        // widening — the connector never invents a codec or a rate.
        const format: AssistantAudioFormat | undefined =
          (encoding === "opus" || encoding === "pcm") && typeof sampleRate === "number"
            ? { encoding, sampleRate }
            : undefined;
        log.debug("audio-start", { turnId: this.activeTurnId, encoding, sampleRate });
        this.config.onAudioStart?.(this.activeTurnId, format);
      }),
    );

    this.unsubs.push(
      sdk.onBinary((data: ArrayBuffer) => {
        if (!this.isReceiving || this.isCancelled) return;
        this.config.onAudioFrame?.(new Uint8Array(data), this.activeTurnId);
      }),
    );

    this.unsubs.push(
      sdk.onMessage("turn.audio.done", (msg: unknown) => {
        const m = msg as { turnId?: string };
        const doneTurnId = m.turnId ?? this.activeTurnId;
        this.isReceiving = false;
        log.debug("audio-done", { turnId: doneTurnId, suppressed: this.isCancelled });
        if (!this.isCancelled) this.config.onAudioDone?.(doneTurnId);
      }),
    );

    this.unsubs.push(
      sdk.onMessage("playback.stop", (msg: unknown) => {
        const m = msg as { turnId?: string; reason?: "barge-in" | "interrupt" };
        // Mark the stream dropped. `isReceiving` stays false until a fresh
        // turn.audio.start arrives — the next turn's audio re-enables playback
        // automatically.
        this.isCancelled = true;
        this.isReceiving = false;
        const reason = m.reason ?? "barge-in";
        log.info("playback-stop", { reason, turnId: m.turnId ?? "" });
        this.config.onPlaybackStop?.(reason, m.turnId ?? "");
      }),
    );
  }

  detach(): void {
    for (const unsub of this.unsubs) unsub();
    this.unsubs = [];
    this.isReceiving = false;
    this.isCancelled = false;
    this.activeTurnId = "";
  }

  onCancelled(): void {
    this.isCancelled = true;
    this.isReceiving = false;
    this.activeTurnId = "";
  }
}

File: shared/mobile-sdk/src/commonMain/kotlin/io/sentient/mobilesdk/connectors/AssistantAudioResponseConnector.kt
// ---------------------------------------------------------------------------
// AssistantAudioResponseConnector — receives assistant audio from the gateway.
//
// Mirrors web-sdk's assistant-audio-response-connector.ts VERBATIM:
//   capability = "audio.output"
//   turn.audio.start → activeTurnId = turnId; isReceiving = true;
//                           isCancelled = false; onAudioStart(turnId)
//   binary frame (handleBinary) → IF isReceiving && !isCancelled →
//                           onAudioFrame(bytes, activeTurnId); else DROP
//   turn.audio.done → isReceiving = false; IF !isCancelled →
//                           onAudioDone(turnId, falling back to activeTurnId when blank)
//   playback.stop → isCancelled = true; isReceiving = false;
//                           onPlaybackStop(reason, defaulting to "barge-in" when blank)
//
// TURN ATTRIBUTION: binary downlink frames carry NO turnId (9-byte header = seq +
// type only), so every frame is attributed to the most recent turn.audio.start.
// The gateway therefore MUST bracket each turn's audio strictly —
// start(t) … frames(t) … done(t) — before start(t+1). AudioPipeline's queue is
// defensive against an overlapping start, but it cannot re-attribute bytes.
//
// DROP-GUARD (the barge-in / interrupt latch — ported exactly):
//   A playback.stop sets isCancelled and clears isReceiving. Subsequent binary
//   frames are DROPPED (the && !isCancelled check) and a trailing
//   turn.audio.done is suppressed — until the NEXT turn.audio.start
//   clears isCancelled and re-enables playback. This is what stops the user
//   hearing the assistant immediately on barge-in while the next turn's audio
//   resumes cleanly.
//
// web-sdk receives binary via sdk.onBinary; mobile-sdk receives it through the
// router calling handleBinary (this connector is MessageRouter.audioConnector).
// web-sdk's separate onCancelled() effect-cancel hook collapses into the SAME
// latch path as playback.stop here — mobile-sdk drives the latch through the
// playback.stop frame, which is the wire signal for barge-in / interrupt.
//
// No coroutines here — the E3 pipeline routes onAudioFrame → VoicePlaybackSink
// .playFrame and onPlaybackStop → .flushPlayback. This connector owns ONLY the flag FSM.
//
// Threading: single-threaded; the router drives handle/handleBinary on the
// orchestrator's dispatcher. The flag/turnId state is owned here.
// ---------------------------------------------------------------------------
package io.sentient.mobilesdk.connectors

import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.protocol.ServerMessage

// JUSTIFIED DIVERGENCE from web-sdk's assistant-audio-response-connector.ts:
// onAudioStart carries encoding + sampleRate. On web, opus→PCM decode lives in
// webui (not web-sdk), so the web connector never sees the encoding. On mobile
// the SDK OWNS decode (AudioPipeline → OpusDecoderPort), so the connector must
// forward turn.audio.start's encoding/sampleRate down to the pipeline.
class AssistantAudioResponseConnector(
    private val onAudioStart: ((turnId: String, encoding: String?, sampleRate: Int?) -> Unit)? = null,
    private val onAudioFrame: ((frame: ByteArray, turnId: String) -> Unit)? = null,
    private val onAudioDone: ((turnId: String) -> Unit)? = null,
    private val onPlaybackStop: ((reason: String, turnId: String) -> Unit)? = null,
) : Connector {
    override val capability: String = CAPABILITY

    private val log = createLogger("connector", "assistant-audio-response")

    /** True between audio.start and audio.done / playback.stop. */
    private var isReceiving = false

    /** Drop latch: set on playback.stop, cleared on the next audio.start. */
    private var isCancelled = false

    /** turnId of the currently active audio stream (set on audio.start). */
    private var activeTurnId = ""

    /** Exposed for tests / orchestration: are we accepting downlink frames? */
    fun isReceiving(): Boolean = isReceiving

    /** Exposed for tests / orchestration: is the current stream cancel-latched? */
    fun isCancelled(): Boolean = isCancelled

    override fun handle(msg: ServerMessage) {
        when (msg) {
            is ServerMessage.TurnAudioStart -> onStart(msg.turnId, msg.encoding, msg.sampleRate)
            is ServerMessage.TurnAudioDone -> onDone(msg.turnId)
            is ServerMessage.PlaybackStop -> onStop(msg.reason, msg.turnId)
            else -> Unit // not owned by this connector
        }
    }

    /** Raw PCM16 LE downlink. Gated by the receiving && !cancelled drop-guard. */
    override fun handleBinary(bytes: ByteArray) {
        if (!isReceiving || isCancelled) {
            log.debug(
                "frame-dropped",
                mapOf("bytes" to bytes.size, "isReceiving" to isReceiving, "isCancelled" to isCancelled),
            )
            return
        }
        log.debug("downlink-frame", mapOf("bytes" to bytes.size, "turnId" to activeTurnId))
        onAudioFrame?.invoke(bytes, activeTurnId)
    }

    private fun onStart(turnId: String, encoding: String?, sampleRate: Int?) {
        activeTurnId = turnId
        isReceiving = true
        isCancelled = false
        log.info(
            "transition",
            mapOf(
                "to" to "receiving",
                "trigger" to "turn.audio.start",
                "turnId" to turnId,
                "encoding" to (encoding ?: ""),
                "sampleRate" to (sampleRate ?: 0),
            ),
        )
        onAudioStart?.invoke(turnId, encoding, sampleRate)
    }

    private fun onDone(turnId: String) {
        // Blank turnId (a malformed/legacy frame) falls back to the stream in flight.
        val doneId = turnId.ifEmpty { activeTurnId }
        isReceiving = false
        log.info(
            "transition",
            mapOf(
                "to" to "idle",
                "trigger" to "turn.audio.done",
                "turnId" to doneId,
                "isCancelled" to isCancelled,
            ),
        )
        if (!isCancelled) onAudioDone?.invoke(doneId)
    }

    private fun onStop(reason: String, turnId: String) {
        // Mark this stream dropped. isReceiving stays false until a fresh
        // turn.audio.start arrives — the next turn's audio re-enables
        // playback automatically.
        isCancelled = true
        isReceiving = false
        val effectiveReason = reason.ifEmpty { DEFAULT_STOP_REASON }
        log.info(
            "transition",
            mapOf(
                "to" to "cancelled",
                "trigger" to "playback.stop",
                "reason" to effectiveReason,
                "turnId" to turnId,
            ),
        )
        onPlaybackStop?.invoke(effectiveReason, turnId)
    }

    companion object {
        const val CAPABILITY: String = "audio.output"

        /** web-sdk defaults an absent playback.stop reason to "barge-in". */
        private const val DEFAULT_STOP_REASON: String = "barge-in"
    }
}
